From 74781b6715a48f91bff022f96cf1d006612f3e4b Mon Sep 17 00:00:00 2001
From: Ferry Toth <ftoth@exalondelft.nl>
Date: Tue, 18 May 2021 16:53:22 +0200
Subject: [PATCH] serial: 8250_port: when using DMA do not split writes in TTY

By default tty_io write() splits large writes into 2K chunks. This
results to a single DMA transfer to split into 2 seperate transfers
which are not combined on the scatter gather list. As a result, under
heavy load or PM conditions an interchar gap occurs which can lead
to errors in the receiver.

Signed-off-by: Ferry Toth <ftoth@exalondelft.nl>
---
 drivers/tty/serial/8250/8250_port.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/tty/serial/8250/8250_port.c b/drivers/tty/serial/8250/8250_port.c
index b0af13074cd3..6781fef254dd 100644
--- a/drivers/tty/serial/8250/8250_port.c
+++ b/drivers/tty/serial/8250/8250_port.c
@@ -2397,6 +2397,13 @@ int serial8250_do_startup(struct uart_port *port)
 		if (msg) {
 			dev_warn_ratelimited(port->dev, "%s\n", msg);
 			up->dma = NULL;
+		} else {
+			/*
+			 * change tty_io write() to not split
+			 * large writes into 2K chunks
+			 */
+			set_bit(TTY_NO_WRITE_SPLIT,
+				&port->state->port.tty->flags);
 		}
 	}
 
-- 
2.30.2

